File: codes/algo/ufs/ufs.go
package main

type Ufs struct {
	p         []int
	cnt       []int
	component int
}

func NewUfs(size int) *Ufs {
	p := make([]int, size)
	cnt := make([]int, size)
	for i := range p {
		p[i] = i
		cnt[i] = 1
	}
	return &Ufs{p, cnt, size}
}

func (u *Ufs) union(a, b int) {
	x, y := u.find(a), u.find(b)
	if x == y {
		return
	}
	rankX, rankY := u.cnt[x], u.cnt[y]
	if rankX <= rankY {
		u.p[x] = y
		u.cnt[y] += u.cnt[x]
	} else {
		u.p[y] = x
		u.cnt[x] += u.cnt[y]
	}
	u.component--
	return
}

func (u *Ufs) find(a int) int {
	if u.p[a] != a {
		u.p[a] = u.find(u.p[a])
	}
	return u.p[a]
}

File: codes/algo/ufs/ufs.py
# union found set impl
class Ufs:
    def __init__(self, size):
        self.parent = list(range(size))  # 初始化每个节点的父节点为自己

    def union(self, a, b) -> bool:
        """并 合并为一个集合"""
        root_a, root_b = self.find(a), self.find(b)
        if root_a == root_b:
            return True
        self.parent[root_a] = root_b
        return False

    def find(self, a):  # 路径压缩，不完全的
        """查a的根节点"""
        p = self.parent[a]
        while p != self.parent[p]:
            p = self.parent[p]
        self.parent[a] = p
        return p
